### PR TITLE
Feature configurable mgmt sim scenario

### PR DESCRIPTION
**Pull request by kaizimmerm**
I made the Management API simulator scenario configurable (e.g. number of targets, DS, etc.) which allows to use this together with device simulator for a complete roundtrip.

I used this also for a couple of load tests on the DMF API which resulted in a couple of improvements I did there:
* DMF AMQP has not multiple consumers and a higher than 1 prefetch count.
* That improved throughput a lot but comes with a price. Message coming in have to guaranteed order of processing anymore. The only area with this might be problematic is the action feedback channel, i.e. message close message is processed before some of the intermediate feedback. I introduced a mechanism to allow late feedback.
* All of this stuff is configurable so users can go back to the old mechanism.

In addition I fixed a few minor issues that came along the way while working with mgmt and DMF APIs:
* Fixed wrong parameter in target REST API (controllerId instead of targetId).
* I teached the device simulator to auto poll the same way for DMF as it already did for DDI
* I fixed a problematic exception handling in DMF where all message where send to dead letter instead of requeued in case of internal errors.
* I added the option to set the amqp address on targets by management API.

### CODE REVIEW

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaControllerManagement.java [hunk 1]
```diff
@@ -249,14 +253,26 @@ private void handleFinishedCancelation(final ActionStatus actionStatus, final Jp
     public Action addUpdateActionStatus(@NotNull final ActionStatus actionStatus) {
         final JpaAction action = (JpaAction) actionStatus.getAction();
 
-        if (!action.isActive()) {
+        // TODO: test
```
Comment (michahirsch) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaControllerManagement.java]: check this todo!

**Comment by michahirsch** [comment]
I've added the `InvalidTargetAddressException` in case the given address from a user violates the RFC URI standard a http-response code 400 (bad request) is then returned instead of 500 (internal server error)

File: examples/hawkbit-example-mgmt-simulator/src/main/java/org/eclipse/hawkbit/mgmt/client/scenarios/ConfigurableScenario.java [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+        // put the next byte in the array
+        random.nextBytes(nbyte);
+
+        return nbyte;
+    }
+
+    private void createTargets(final Scenario scenario) {
+        LOGGER.info("Creating {} targets", scenario.getTargets());
+
+        for (int i = 0; i < scenario.getTargets() / 100; i++) {
```
Comment (schabdo) [inline examples/hawkbit-example-mgmt-simulator/src/main/java/org/eclipse/hawkbit/mgmt/client/scenarios/ConfigurableScenario.java]: Introduce brackets `(scenario.getTargets() / 100)`

File: examples/hawkbit-example-mgmt-simulator/src/main/java/org/eclipse/hawkbit/mgmt/client/scenarios/ConfigurableScenario.java [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+        // TODO: complete this as soon as rollouts can be deleted
+
+    }
+
+    private void deleteSoftwareModules() {
+        PagedList<MgmtSoftwareModule> modules;
+        do {
+            modules = softwareModuleResource.getSoftwareModules(0, 100, null, null).getBody();
+            modules.getContent().forEach(module -> softwareModuleResource.deleteSoftwareModule(module.getModuleId()));
+        } while (modules.getTotal() > 100);
```
Comment (schabdo) [inline examples/hawkbit-example-mgmt-simulator/src/main/java/org/eclipse/hawkbit/mgmt/client/scenarios/ConfigurableScenario.java]: `100` is used very often. Extract magic number to constant

File: examples/hawkbit-example-mgmt-simulator/src/main/java/org/eclipse/hawkbit/mgmt/client/scenarios/upload/ArtifactFile.java [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+     * @param contentType
+     *            the content type
+     * @param content
+     *            of the file
+     */
+    public ArtifactFile(final String name, final String originalFilename, final String contentType,
+            final byte[] content) {
+        Assert.hasLength(name, "Name must not be null");
+        this.name = name;
+        this.originalFilename = originalFilename != null ? originalFilename : "";
```
Comment (schabdo) [inline examples/hawkbit-example-mgmt-simulator/src/main/java/org/eclipse/hawkbit/mgmt/client/scenarios/upload/ArtifactFile.java]: Mybe use: Optional.ofNullable(originalFilename).orElse("")

File: examples/hawkbit-example-mgmt-simulator/src/main/java/org/eclipse/hawkbit/mgmt/client/scenarios/upload/FeignMultipartEncoder.java [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+        }
+
+        @Override
+        public String getFilename() {
+            return this.filename;
+        }
+
+        @Override
+        public InputStream getInputStream() throws IOException, IllegalStateException {
+            return super.getInputStream(); // To change body of generated
```
Comment (schabdo) [inline examples/hawkbit-example-mgmt-simulator/src/main/java/org/eclipse/hawkbit/mgmt/client/scenarios/upload/FeignMultipartEncoder.java]: What does this comment mean?

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaControllerManagement.java [hunk 6]
```diff
@@ -249,14 +253,25 @@ private void handleFinishedCancelation(final ActionStatus actionStatus, final Jp
     public Action addUpdateActionStatus(@NotNull final ActionStatus actionStatus) {
         final JpaAction action = (JpaAction) actionStatus.getAction();
 
-        if (!action.isActive()) {
+        // if action is already closed we accept further status updates on if
```
Comment (schabdo) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaControllerManagement.java]: remove "on"

File: hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaControllerManagement.java [hunk 7]
```diff
@@ -249,14 +253,25 @@ private void handleFinishedCancelation(final ActionStatus actionStatus, final Jp
...
+        // permitted so by configuration. This is especially use full if the
```
Comment (schabdo) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaControllerManagement.java]: Typo => usefull

Comment (kaizimmerm) [inline hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaControllerManagement.java]: Its actually "useful" 😄

**Pull request merged**